src/gensxg.c

```diff
@@ -1,128 +1,217 @@
+    "-expire string\n"
+    "  The expire time of the SXG in RFC3339 format (2006-01-02T15:04:05Z)."
+    " (default <date> +7 days)\n";
 
 typedef struct {
   bool help;
+  bool integrity_mode;
+  const char* content;
+  const char* content_type;
+  sxg_header_t header;
+  uint64_t mi_record_size;
   const char* url;
   const char* cert_url;
   const char* validity_url;
   const char* certificate;
   const char* public_key;
   const char* private_key;
-  const char* content;
-  const char* content_type;
-  uint64_t date;
-  uint64_t expire;
-  uint64_t mi_record_size;
+  time_t date;
+  time_t expire;
 } Options;
 
 static EVP_PKEY* load_private_key(const char* filepath) {
   FILE* const keyfile = fopen(filepath, "r");
   if (keyfile == NULL) {
-    return NULL;
+    fprintf(stderr, "fopen %s: %s\n", strerror(errno), filepath);
+    exit(EXIT_FAILURE);
   }
   EVP_PKEY* private_key =
       PEM_read_PrivateKey(keyfile, NULL, NULL, NULL);
+  if (private_key == NULL) {
+    fprintf(stderr, "Failed to read private key: %s\n", filepath);
+    exit(EXIT_FAILURE);
+  }
   fclose(keyfile);
   return private_key;
 }
 
 static X509* load_x509_cert(const char* filepath) {
   FILE* certfile = fopen(filepath, "r");
   if (certfile == NULL) {
-    return NULL;
+    fprintf(stderr, "fopen %s: %s\n", strerror(errno), filepath);
+    exit(EXIT_FAILURE);
   }
   X509* cert = PEM_read_X509(certfile, NULL, NULL, NULL);
+  if (cert == NULL) {
+    fprintf(stderr, "Failed to read cetificate: %s\n", filepath);
+    exit(EXIT_FAILURE);
+  }
   fclose(certfile);
   return cert;
 }
 
 static EVP_PKEY* load_ed25519_pubkey(const char* filepath) {
   FILE* keyfile = fopen(filepath, "r");
-  if (!keyfile) {
-    return NULL;
+  if (keyfile == NULL) {
+    fprintf(stderr, "fopen %s: %s\n", strerror(errno), filepath);
+    exit(EXIT_FAILURE);
   }
   EVP_PKEY* public_key = PEM_read_PUBKEY(keyfile, NULL, NULL, NULL);
+  if (public_key == NULL) {
+    fprintf(stderr, "Failed to read public key: %s\n", filepath);
+    exit(EXIT_FAILURE);
+  }
   fclose(keyfile);
   return public_key;
 }
 
-Options parse_options(int argc, char* const argv[]) {
+static time_t parse_time(const char datestring[]) {
+  struct tm date;
+  char* finished = strptime(datestring, "%Y-%m-%dT%H:%M:%SZ", &date);
+  if (finished == NULL) {
+    fprintf(stderr, "Failed to parse date: %s\n", datestring);
+  } else if(*finished != '\0') {
+    fprintf(stderr, "Failed to parse at: %s\n", finished);
+    exit(EXIT_FAILURE);
+  }
+  return mktime(&date);
+}
+
+Options init_default_options() {
   Options result;
   memset(&result, 0, sizeof(result));
-  // Set default values.
   result.help = false;
+  result.integrity_mode = false;
+  result.content = "index.html";
   result.content_type = "text/html";
+  result.header = sxg_empty_header();
   time_t now = time(NULL);
-  result.date = (uint64_t)now;
-  result.expire = result.date + 60 * 60 * 24 * 7;  // 7 days.
+  result.date = now;
```

Comment:
Moving out the init of `expires` makes the code structure messed up (though I understand the necessity), I'd suggest holding `duration` instead of `expires` in `Options` struct.

Reply:
I'm not sure why having `duration` solves the problem drastically. Please add more explanation.

Reply:
I don't think it will be "drastic". Something like `set_expire` will be still necessary in `parse_options` but we can state all the default values in `init_default_options`, which I meant by "code structure".

Reply:
OK, I understand. But I think giving duration in string format in C will introduce another complexity. I'll append `set_expire` function to uniformity of code structure.

Reply:
I meat like:
```
struct Options {
  ...
  time_t date;
  SOME_TYPE_REPRESENTING_TIME_DIFF duration;
}

// Now we can include all the "defaults" here.
Options init_default_options() {
  ...
  result.date = time(NULL);
  result.duration = <7 days>;
}

Options parse_options(...) {
  bool is_expire_specified = false;
  ...
    case CASE_FOR_EXPIRE:
      is_expire_specified = true;
      <remember the specified value>;
  ...
  if (is_expire_specified) {
    <update duration based on that value>;
  }
}
```

Anyway I don't think this is a big deal so let me keep this resolved.

Reply:
OK, using duration.